ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/SearchIndexTool.java

```diff
@@ -0,0 +1,195 @@
+
+    private String name = TYPE;
+
+    private String description = DEFAULT_DESCRIPTION;
+
+    private Client client;
+
+    private NamedXContentRegistry xContentRegistry;
+
+    public static final Gson gson;
+
+    static {
+        gson = new Gson();
+    }
+
+    public SearchIndexTool(Client client, NamedXContentRegistry xContentRegistry) {
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    public String getType() {
+        return TYPE;
+    }
+
+    @Override
+    public String getVersion() {
+        return null;
+    }
+
+    @Override
+    public boolean validate(Map<String, String> parameters) {
+        if (parameters == null || !parameters.containsKey(INPUT_FIELD)) {
+            return false;
+        }
+        String input = parameters.get(INPUT_FIELD);
+        return input != null;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        try {
+            String input = parameters.get(INPUT_FIELD);
+            String index = "";
+            String query = "";
+            try {
+                // need this try catch block, because index, query field may not exist
+                JsonObject jsonObject = gson.fromJson(input, JsonObject.class);
+                index = jsonObject.get(INDEX_FIELD).getAsString();
+                query = jsonObject.get(QUERY_FIELD).toString();
+                query = "{\"query\": " + query + "}";
+            } catch (Exception e) {
+                // throw new IllegalArgumentException("wrong input");
+            }
+            if (org.apache.commons.lang3.StringUtils.isBlank(query)) {
+                throw new IllegalArgumentException("[" + QUERY_FIELD + "] is null or empty, can not process it.");
+            }
+
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            XContentParser queryParser = XContentType.JSON
+                .xContent()
+                .createParser(xContentRegistry, LoggingDeprecationHandler.INSTANCE, query);
+            searchSourceBuilder.parseXContent(queryParser);
+            SearchRequest searchRequest = new SearchRequest().source(searchSourceBuilder).indices(index);
+
+            ActionListener actionListener = ActionListener.<SearchResponse>wrap(r -> {
+                SearchHit[] hits = r.getHits().getHits();
+
+                if (hits != null && hits.length > 0) {
+                    StringBuilder contextBuilder = new StringBuilder();
+                    for (int i = 0; i < hits.length; i++) {
+                        SearchHit hit = hits[i];
+                        String doc = AccessController.doPrivileged((PrivilegedExceptionAction<String>) () -> {
+                            Map<String, Object> docContent = new HashMap<>();
+                            docContent.put("_index", hit.getIndex());
```

Comment:
Why extracting these fields instead of returning the `SearchResponse` directly?

Reply:
I will return SearchResponse directly.

Reply:
The returned response should be `StringBuilder contextBuilder`.

Reply:
We can return the processed results from the search response.